src/libPersistence/ContractStorage2.cpp

```diff
@@ -0,0 +1,968 @@
+
+  for (unsigned int i = 2; i < indices.size() - 1 /*exclude the value key*/;
+       ++i) {
+    scankey += indices.at(i) + SCILLA_INDEX_SEPARATOR;
+    DeleteByIndex(scankey);
+  }
+  return true;
+}
+
+void ContractStorage2::UpdateStateData(const string& key, const bytes& value,
+                                       bool cleanEmpty) {
+  if (LOG_SC) {
+    LOG_GENERAL(INFO, "key: " << key << " value: "
+                              << DataConversion::CharArrayToString(value));
+  }
+
+  if (cleanEmpty) {
+    CleanEmptyMapPlaceholders(key);
+  }
+
+  auto pos = m_indexToBeDeleted.find(key);
+  if (pos != m_indexToBeDeleted.end()) {
+    m_indexToBeDeleted.erase(pos);
+  }
+
+  t_stateDataMap[key] = value;
+}
+
+bool ContractStorage2::UpdateStateValue(const dev::h160& addr, const bytes& q,
+                                        unsigned int q_offset, const bytes& v,
+                                        unsigned int v_offset) {
+  if (LOG_SC) {
+    LOG_MARKER();
+  }
+
+  if (q_offset > q.size()) {
+    LOG_GENERAL(WARNING, "Invalid query data and offset, data size "
+                             << q.size() << ", offset " << q_offset);
+    return false;
+  }
+
+  if (v_offset > v.size()) {
+    LOG_GENERAL(WARNING, "Invalid value data and offset, data size "
+                             << v.size() << ", offset " << v_offset);
+  }
+
+  ProtoScillaQuery query;
+  query.ParseFromArray(q.data() + q_offset, q.size() - q_offset);
+
+  if (!query.IsInitialized()) {
+    LOG_GENERAL(WARNING, "Parse bytes into ProtoScillaQuery failed");
+    return false;
+  }
+
+  ProtoScillaVal value;
+  value.ParseFromArray(v.data() + v_offset, v.size() - v_offset);
+
+  if (!value.IsInitialized()) {
+    LOG_GENERAL(WARNING, "Parse bytes into ProtoScillaVal failed");
+    return false;
+  }
+
+  string key = addr.hex() + SCILLA_INDEX_SEPARATOR + query.name() +
+               SCILLA_INDEX_SEPARATOR;
+
+  if (query.ignoreval()) {
+    if (query.indices().size() < 1) {
+      LOG_GENERAL(WARNING, "indices cannot be empty")
+      return false;
+    }
+    for (int i = 0; i < query.indices().size() - 1; ++i) {
+      key += query.indices().Get(i) + SCILLA_INDEX_SEPARATOR;
+    }
+    string parent_key = key;
+    key += query.indices().Get(query.indices().size() - 1) +
+           SCILLA_INDEX_SEPARATOR;
+    if (LOG_SC) {
+      LOG_GENERAL(INFO, "Delete key: " << key);
+    }
+    DeleteByPrefix(key);
+
+    map<string, bytes> t_states;
+    FetchStateDataForKey(t_states, parent_key, true);
+    if (t_states.empty()) {
+      ProtoScillaVal empty_val;
+      empty_val.mutable_mval()->mutable_m();
+      bytes dst;
+      if (!SerializeToArray(empty_val, dst, 0)) {
+        LOG_GENERAL(WARNING, "empty_mval SerializeToArray failed");
+        return false;
+      }
+      UpdateStateData(parent_key, dst);
+    }
+  } else {
+    for (const auto& index : query.indices()) {
+      key += index + SCILLA_INDEX_SEPARATOR;
+    }
```

Comment:
I think it's good to assert that `query.name() != FIELDS_MAP_DEPTH_INDICATOR`, and also assert that none of the indices contain `SCILLA_INDEX_SEPARATOR`.

The comment applies to `FetchStateValue` also.

Reply:
Added, good point